improver_tests/calibration/reliability_calibration/test_ManipulateReliabilityTable.py

```diff
@@ -30,659 +30,533 @@
+
+    expected = np.array(
+        [
+            [0, 262, 325],  # Observation count
+            [0, 262, 325],  # Sum of forecast probability
+            [1000, 550, 350],  # Forecast count
+        ]
+    )
+    result = Plugin()._combine_undersampled_bins(
+        obs_count, forecast_probability_sum, forecast_count, probability_bin_coord,
+    )
+
+    assert_array_equal(result[:3], expected)
+    expected_bin_coord_points = np.array([0.1, 0.4, 0.8], dtype=np.float32)
+    expected_bin_coord_bounds = np.array(
+        [[0.0, 0.2], [0.2, 0.6], [0.6, 1.0]], dtype=np.float32,
+    )
+    assert_allclose(expected_bin_coord_points, result[3].points)
+    assert_allclose(expected_bin_coord_bounds, result[3].bounds)
+
+
+def test_cub_two_equal_undersampled_bins(probability_bin_coord):
+    """Test when two bins are under-sampled and the under-sampled bins have
+    an equal forecast count."""
+    obs_count = np.array([0, 25, 250, 75, 250], dtype=np.float32)
+    forecast_probability_sum = np.array([0, 25, 250, 75, 250], dtype=np.float32)
+    forecast_count = np.array([1000, 100, 500, 100, 250], dtype=np.float32)
+
+    expected = np.array(
+        [
+            [0, 275, 325],  # Observation count
+            [0, 275, 325],  # Sum of forecast probability
+            [1000, 600, 350],  # Forecast count
+        ]
+    )
+
+    result = Plugin()._combine_undersampled_bins(
+        obs_count, forecast_probability_sum, forecast_count, probability_bin_coord,
+    )
+
+    assert_array_equal(result[:3], expected)
+    expected_bin_coord_points = np.array([0.1, 0.4, 0.8], dtype=np.float32)
+    expected_bin_coord_bounds = np.array(
+        [[0.0, 0.2], [0.2, 0.6], [0.6, 1.0]], dtype=np.float32,
+    )
+    assert_allclose(expected_bin_coord_points, result[3].points)
+    assert_allclose(expected_bin_coord_bounds, result[3].bounds)
+
+
+def test_cub_three_equal_undersampled_bin_neighbours(probability_bin_coord):
+    """Test when three neighbouring bins are under-sampled."""
+    obs_count = np.array([0, 25, 50, 75, 250], dtype=np.float32)
+    forecast_probability_sum = np.array([0, 25, 50, 75, 250], dtype=np.float32)
+    forecast_count = np.array([1000, 100, 100, 100, 250], dtype=np.float32)
+
+    expected = np.array(
+        [
+            [0, 150, 250],  # Observation count
+            [0, 150, 250],  # Sum of forecast probability
+            [1000, 300, 250],  # Forecast count
+        ]
+    )
+
+    result = Plugin()._combine_undersampled_bins(
+        obs_count, forecast_probability_sum, forecast_count, probability_bin_coord,
+    )
+
+    assert_array_equal(result[:3], expected)
+    expected_bin_coord_points = np.array([0.1, 0.5, 0.9], dtype=np.float32)
+    expected_bin_coord_bounds = np.array(
+        [[0.0, 0.2], [0.2, 0.8], [0.8, 1.0]], dtype=np.float32,
+    )
+    assert_allclose(expected_bin_coord_points, result[3].points)
+    assert_allclose(expected_bin_coord_bounds, result[3].bounds)
+
+
```

Comment:
We've lost the null case (called `test_monotonic` prior to this refactor) here where cbp doesn't have to do anything. Any particular reason this test was dropped?

Reply:
Good pick up.
I think that test:
```
    def test_monotonic(self):
        """Test no bin pairs are combined, if all bin pairs are monotonic."""
        obs_count = np.array([0, 250, 500, 750, 1000], dtype=np.float32)
        result = Plugin()._combine_bin_pair(
            obs_count,
            self.forecast_probability_sum,
            self.forecast_count,
            self.probability_bin_coord,
        )
        assert_array_equal(
            result[:3], [obs_count, self.forecast_probability_sum, self.forecast_count],
        )
        self.assertEqual(result[3], self.probability_bin_coord)
```
is exactly the same as this test:
```
    def test_no_undersampled_bins(self):
        """Test no bins are combined when no bins are under-sampled."""
        forecast_count = np.array([1000, 1000, 1000, 1000, 1000], dtype=np.float32)

        result = self.plugin._combine_undersampled_bins(
            self.obs_count,
            self.forecast_probability_sum,
            forecast_count,
            self.probability_bin_coord,
        )

        assert_array_equal(
            result[:3], [self.obs_count, self.forecast_probability_sum, forecast_count]
        )
        self.assertEqual(result[3], self.probability_bin_coord)
```
It is not easily obvious due to convoluted input setup. Thus I think I updated the docstring of the first test and removed the duplicate.

If you are happy do you mind approving? Can't merge otherwise, thanks

Reply:
The `test_monotonic`, was checking that[ this if statement](https://github.com/metoppv/improver/blob/accc6133769d5b326bbebdf463a2ea82f757ab57/improver/calibration/reliability_calibration.py#L839) is false in the case that the observation frequency is monotonic across bins. The `test_no_undersampled_bins` test checks that [this while statement](https://github.com/metoppv/improver/blob/accc6133769d5b326bbebdf463a2ea82f757ab57/improver/calibration/reliability_calibration.py#L764) is false and nothing is done in the case that there are sufficient forecast counts in each bin. I was concerned we were losing the coverage of the former function returning unchanged inputs, however I think [this line](https://github.com/metoppv/improver/blob/accc6133769d5b326bbebdf463a2ea82f757ab57/improver/calibration/reliability_calibration.py#L994) means we can never access the function in such a case, so I am happy that the test is unnecessary.

Reply:
Whoops, I was so focused on the inputs I totally forgot that they are testing different function. If you think it's unnecessary, I'll leave as is, thanks!